### PR TITLE
[corrade,magnum,-plugins,-extras,-integration] Update to latest and support feature packages

### PR DESCRIPTION
**Pull request by Squareys**
Hi everybody!

I update the corrade package, added support for feature packages and updated the project website URL.

Once that is through, will create another pullrequest for magnum/magnum-plugins/magnum-extras and magnum-integration packages. (In case someone suggests an improvement here, it wont be replicated all over those packages).

@mosra Could you give this a quick review? Thanks! :)

Cheers, Jonathan.

### CODE REVIEW

**Comment by mosra** [comment]
@Squareys Thank you. :+1: 

Feature packages: with the current state of Corrade (where everything is `ON` by default) the feature packages as they are implemented now won't do any difference, so `corrade[interconnect]` will build the same stuff as `corrade` and there's nothing more to "opt in". However, it would be nice to "opt out". I searched around a bit and found that there's [Default-Features](https://github.com/Microsoft/vcpkg/blob/master/docs/specifications/feature-packages.md#signalrclient) which lists the things that are enabled by default when installing just `corrade`. Then there could be some logic to enable what is used and disable what's not in case one installs for example only `corrade[pluginmanager]`. That should also correctly handle inter-feature dependencies.

I guess you'll also need to take care of removing duplicate files when installing just a particular feature, as some parts are installed always (for example CMake scripts).

**Comment by ras0219-msft** [comment]
@mosra thanks for digging into the docs! Unfortunately, default features aren't quite implemented yet.

**Comment by Squareys** [comment]
@ras0219-msft How will that be handled then? Will it only enable `[core]` by default?

**Comment by ras0219-msft** [comment]
Yes, that's correct.

Also, dependencies on features do work correctly, so those can be used without concern.

**Comment by Squareys** [comment]
@ras0219-msft Is someone already working on the "Default-Features"? Otherwise I'd give it a shoot.

**Comment by ras0219-msft** [comment]
It's on the backlog, but we're not actively attacking it at the moment.

I'll warn you up front that it might be challenging, but I'll be happy to provide feedback and I'd love to have more eyes on the codebase regardless :)! Due to the complexity of the feature, I think it will be easiest to handle by a TDD approach.

**Comment by Squareys** [comment]
@ras0219-msft I dug through the code a bit and figured out that `""` feature is handled as core.

For the default features I currently see two options
 (a) handle `""` as default features which turned out to be tricky, as the information about which features are default is not avaible at all of those locations where `""` is handled.
 (b) introduce "default" alongside "core", which could be represented as a feature in the Cluster (?) for a package spec with a dependency/edge to all the default packages. This "default" feature would be added in dependencies.cpp#L87 `cluster_from_scf`.

I believe (b) let's be easily change code that replaces "" with "core" to replace it with "default" instead. I wonder how that would be represented to the user, though.
Do you see any problems with (b) ? Otherwise I'll continue to try to figure out (a) :D
Any other random hints?

Thanks in advance! (I will open a WIP PR shortly so we can move the discussion there)

**Comment by ras0219-msft** [comment]
I don't believe default can be handled (directly) as an individual feature. Let me lay out a few use cases that may demonstrate the complexity:

```
>vcpkg install z0 // z0 depends on x0[core]
// should install defaults of x0 and z0

>vcpkg install x1[core,y]
>vcpkg install z1 // z1 depends on x1[core]
// should not install defaults of x1, but should install defaults of z1

>vcpkg install z2 x2[core,y] // z2 depends on x2[core]
// should not install defaults of x2, but should install defaults of z2

>vcpkg install x3[core,y]
>vcpkg upgrade
// should not install defaults of x3

>vcpkg install z4 x4 // z4 depends on x4[core]
// should install defaults of both x4 and z4

>vcpkg install z5 // z5 depends on x5[core]
// should install defaults of both z5 and x5 <- this is hard

>vcpkg install z6[core,w] // z6 depends on x6[core]
// should not install defaults of z6 but should install defaults of x6

>vcpkg install x7[core,y]
>vcpkg install x7
// should install defaults of x7
```
The really tricky behavior is that a reference somewhere down the list on the command line can _disable_ installing the defaults, but otherwise the defaults will be installed.

My intuition is that the easiest way to implement it will be to gather a list upfront of all packages that are named on the command line in a way that will disable defaults and pass that list down through the graph algorithm. When a package is being installed "fresh" (i.e. isn't currently installed), and that package is not in the "list-of-defaults-disabled", then the default list is consulted and installed.

If possible, I'd also like to avoid cluttering up the list of installed features with a dummy "defaults" feature unless it is needed to make the algorithm significantly cleaner.

**Comment by Squareys** [comment]
~~~
>vcpkg install z0 // z0 depends on x0[core]
// should install defaults of x0 and z0 <- this the hard part
~~~
If it explicitly depends on x0[core] it should only be installing the core, not the defaults, right?
I believe in this scenario it should install x0[core] and the defaults of z0.

~~~
>vcpkg install z6[core,w] // z6 depends on x6[core]
// should not install defaults of z6 but should install defaults of x6
~~~
Again, my current mental model would require this to install no defaults at all, as [core] is specified explicitly, which removes all the defaults.

From the ["Proposed solution"](https://github.com/Microsoft/vcpkg/blob/master/docs/specifications/feature-packages.md#3-proposed-solution):

> When using vcpkg install <package>, some features will be enabled by default. These default features can be avoided by referring to the packages as <package>[core] and features can be added by supplying them on the same installation line.

So I believe with `[core]` no default features should be added. I will push my current working state and open a WIP PR. I feel like either I am overseeing something important (e.g. misunderstood how the clusters handle inter-feature dependencies) or the "default" feature would indeed make the algorithm significantly cleaner.

**Comment by ras0219-msft** [comment]
While we're working on Default-Features in the other PR, do you want to get this merged without them? I'd like to remove the `Default-Features:` line in the CONTROL file, just to make sure it doesn't trip any half-implemented code that might already exist.

**Comment by Squareys** [comment]
@ras0219-msft I should have the time over the weekend to hopefully get an initial implementation of default feautures running. With reviews and some fixes, it could be done until Wednesday, maybe? In which case I would wait with this PR until then. But maybe I'm being overly optimistic?

**Comment by Squareys** [comment]
[This previous discussion](https://github.com/Microsoft/vcpkg/commit/51ba20d3d14a852887807c1bc36e45a9f2956352#r27530444) on a commit got lost because of a rebase.

@mosra Would be great if you could give this a final quick look... I changed quite a few things while testing that the ports compile correctly.

@ras0219-msft After that final review, this will be ready for merge, finally! The other two PRs I opened are not required for this (though "build" command would fail without #2817). Thank you for your patience :)

File: ports/magnum-plugins/CONTROL [hunk 1]
```diff
@@ -1,4 +1,81 @@
...
+
+Feature: assimpimporter
+Description: AssimpImporter plugin
+Build-Depends: assimp, magnum-plugins[anyimageimporter]
+
+Feature: ddsimporter
+Description: DdsImporter plugin
+
+Feature: devilimageimporter
+Description: DevILImageImporter plugin
```
Comment (mosra) [inline ports/magnum-plugins/CONTROL]: Minor: the correctly cased name is `DevIlImageImporter`. Just sayin' so people don't have problems loading the plugin later when copypasting from here :)

File: ports/magnum/CONTROL [hunk 2]
```diff
@@ -1,4 +1,95 @@
...
+
+Feature: glutapplication
+Description: GlutApplication library
+Build-Depends: freeglut
+
+Feature: audio
+Description: Audio library
+Build-Depends: openal-soft
+
+Feature: magnuminfo
```
Comment (mosra) [inline ports/magnum/CONTROL]: I would say just `info` here, for consistency with the rest

File: ports/magnum/portfile.cmake [hunk 3]
```diff
@@ -48,16 +53,22 @@ vcpkg_configure_cmake(
...
 
 # Drop a copy of tools
-file(COPY ${CURRENT_PACKAGES_DIR}/bin/magnum-distancefieldconverter.exe DESTINATION ${CURRENT_PACKAGES_DIR}/tools/magnum)
-file(COPY ${CURRENT_PACKAGES_DIR}/bin/magnum-fontconverter.exe DESTINATION ${CURRENT_PACKAGES_DIR}/tools/magnum)
+if(distancefieldconverter IN_LIST FEATURES)
+    file(COPY ${CURRENT_PACKAGES_DIR}/bin/magnum-distancefieldconverter.exe DESTINATION ${CURRENT_PACKAGES_DIR}/tools/magnum)
+endif()
+if(fontconverter IN_LIST FEATURES)
+    file(COPY ${CURRENT_PACKAGES_DIR}/bin/magnum-fontconverter.exe DESTINATION ${CURRENT_PACKAGES_DIR}/tools/magnum)
+endif()
```
Comment (mosra) [inline ports/magnum/portfile.cmake]: Isn't the `info`, `al-info` missing here?

File: ports/magnum-plugins/portfile.cmake [hunk 4]
```diff
@@ -45,27 +50,23 @@ vcpkg_configure_cmake(
...
-   file(COPY ${LIB_TO_MOVE_DBG} DESTINATION ${CURRENT_PACKAGES_DIR}/debug/lib)
-   file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/lib/magnum)
-else()
-   # remove headers and libs for plugins
-   file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/include)
-   file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/lib)
-   file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/lib)
-   # hint vcpkg
-   set(VCPKG_POLICY_EMPTY_INCLUDE_FOLDER enabled)
-   set(VCPKG_POLICY_DLLS_WITHOUT_LIBS enabled)
```
Comment (mosra) [inline ports/magnum-plugins/portfile.cmake]: Not sure if this shouldn't stay here. See https://github.com/Microsoft/vcpkg/pull/1235 for the original discussion and reasons behind this.

Or maybe I misunderstand what' going on here :sweat_smile:

File: ports/magnum-extras/portfile.cmake [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        -DBUILD_PLUGINS_STATIC=${BUILD_PLUGINS_STATIC}
+        -DMAGNUM_PLUGINS_DEBUG_DIR=${CURRENT_INSTALLED_DIR}/debug/bin/magnum-d
+        -DMAGNUM_PLUGINS_RELEASE_DIR=${CURRENT_INSTALLED_DIR}/bin/magnum
+)
+
+vcpkg_install_cmake()
+
+# Messages to the user
+if("ui" IN_LIST FEATURES)
+    message(WARNING "It is recommended to install one of magnum-plugins[freetypefont,harfbuzzfont,stbtruetypefont] to have the UI library working out of the box")
```
Comment (mosra) [inline ports/magnum-extras/portfile.cmake]: :+1:

File: ports/magnum-extras/portfile.cmake [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    endif()
+endforeach()
+
+vcpkg_configure_cmake(
+    SOURCE_PATH ${SOURCE_PATH}
+    PREFER_NINJA # Disable this option if project cannot be built with Ninja
+    OPTIONS
+        ${_COMPONENT_FLAGS}
+        -DBUILD_STATIC=${BUILD_STATIC}
+        -DBUILD_PLUGINS_STATIC=${BUILD_PLUGINS_STATIC}
```
Comment (mosra) [inline ports/magnum-extras/portfile.cmake]: No such option here.

File: ports/magnum-integration/portfile.cmake [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    endif()
+endforeach()
+
+vcpkg_configure_cmake(
+    SOURCE_PATH ${SOURCE_PATH}
+    PREFER_NINJA # Disable this option if project cannot be built with Ninja
+    OPTIONS
+        ${_COMPONENT_FLAGS}
+        -DBUILD_STATIC=${BUILD_STATIC}
+        -DBUILD_PLUGINS_STATIC=${BUILD_PLUGINS_STATIC}
```
Comment (mosra) [inline ports/magnum-integration/portfile.cmake]: No such option here.

File: ports/magnum-plugins/portfile.cmake [hunk 8]
```diff
@@ -21,22 +21,27 @@ else()
...
-        -DWITH_OPENGEXIMPORTER=ON
-        -DWITH_OPENGEXIMPORTER=ON
-        -DWITH_STANFORDIMPORTER=ON
-        -DWITH_DRWAVAUDIOIMPORTER=ON
-        -DWITH_ANYAUDIOIMPORTER=ON
-        -DWITH_ANYIMAGECONVERTER=ON
-        -DWITH_ANYSCENEIMPORTER=ON
+        ${_COMPONENT_FLAGS}
         -DBUILD_STATIC=${BUILD_STATIC}
         -DBUILD_PLUGINS_STATIC=${BUILD_PLUGINS_STATIC}
```
Comment (mosra) [inline ports/magnum-plugins/portfile.cmake]: No such option here (it's just BUILD_STATIC everywhere, except magnum, where it's BUILD_STATIC and BUILD_PLUGINS_STATIC)

**Review by mosra — COMMENTED** [review]
Thank you for the huge amount of work put into this.

File: ports/magnum-plugins/portfile.cmake (hunk 4 above)
Comment (Squareys) [reply ports/magnum-plugins/portfile.cmake]: Wow, what a discussion 😅 Will revert to how it was before.

File: ports/magnum-integration/portfile.cmake (hunk 7 above)
Comment (Squareys) [reply ports/magnum-integration/portfile.cmake]: That does also include the following two, right?

File: ports/magnum-plugins/portfile.cmake (hunk 4 above)
Comment (mosra) [reply ports/magnum-plugins/portfile.cmake]: Just a reminder: exactly the same should be done for plugins in the magnum repo.

File: ports/magnum-plugins/portfile.cmake (hunk 8 above)
Comment (Squareys) [reply ports/magnum-plugins/portfile.cmake]: Oh, wow, alright 👍

File: ports/magnum-integration/portfile.cmake (hunk 7 above)
Comment (mosra) [reply ports/magnum-integration/portfile.cmake]: That's overriding where the plugins are searched for. I would keep that there, to be safe.

File: ports/magnum-plugins/portfile.cmake (hunk 4 above)
Comment (Squareys) [reply ports/magnum-plugins/portfile.cmake]: Couldn't I change

~~~
        -DMAGNUM_PLUGINS_DEBUG_DIR=${CURRENT_INSTALLED_DIR}/debug/bin/magnum-d
        -DMAGNUM_PLUGINS_RELEASE_DIR=${CURRENT_INSTALLED_DIR}/bin/magnum
~~~

To not install into a subfolder? Or is this just for finding plugins?

File: ports/magnum-plugins/CONTROL (hunk 1 above)
Comment (Squareys) [reply ports/magnum-plugins/CONTROL]: Was copied from https://github.com/mosra/magnum-plugins/blob/master/CMakeLists.txt#L69, you should fix it there too

File: ports/magnum/CONTROL (hunk 2 above)
Comment (Squareys) [reply ports/magnum/CONTROL]: But that breaks my CMake setup 😅

Comment (Squareys) [reply ports/magnum/CONTROL]: To clarify: the cmake option is `WITH_MAGNUMINFO` (https://github.com/mosra/magnum/blob/master/CMakeLists.txt#L78)

As in, all features enable `-DWITH_<feature-lowercase>=ON`

File: ports/magnum/portfile.cmake (hunk 3 above)
Comment (Squareys) [reply ports/magnum/portfile.cmake]: yep

File: ports/magnum/CONTROL (hunk 2 above)
Comment (mosra) [reply ports/magnum/CONTROL]: Argh!

One more thing on my cleanup list. Okay, then leave it as-is. Sorry.

Comment (Squareys) [reply ports/magnum/CONTROL]: I *could* add an exception, but it would be cleaner this way, I think...

File: ports/magnum-plugins/CONTROL (hunk 1 above)
Comment (mosra) [reply ports/magnum-plugins/CONTROL]: Oh. Thanks. Missed that during a review ...

**Comment by Squareys** [comment]
@mosra Done with the requested changes!

**Review by mosra — APPROVED** [review]

**Comment by mosra** [comment]
Oh, also, forgot to mention it here -- I think https://github.com/Microsoft/vcpkg/issues/2564 can be closed now (it was searching for `PngImporter`, which was not available in the previous version).

**Comment by ras0219-msft** [comment]
This all looks great to me!

Thanks both of you for the PR and review, it's great to get direct author feedback 😄!

**Comment by Squareys** [comment]
@ras0219-msft Thanks for the fast merge! Also thanks again for all the guidance with the Default-Features and all!

And @mosra thanks for repeatidly looking at this and for all the feedback 👍

**Pull request merged**